lib/src/facts/linux/operating_system_resolver.cc

```diff
@@ -21,82 +22,186 @@ namespace bs = boost::system;
+        {
+            return _release_info["VERSION_ID"];
         }
+    };
 
-        auto release = get_release(result.name, result.distro.release);
-        if (!release.empty()) {
-            result.release = move(release);
+    struct os_coreos : os_linux
+    {
+        os_coreos(map<string, string> release_info) : os_linux(move(release_info)) {}
+        virtual string get_name(string const& distro_id) override
+        {
+            return os::coreos;
+        }
+        virtual string get_family(string const& name) override
+        {
+            return os_family::coreos;
+        }
+        virtual string get_release(string const& name, string const& distro_release) override
+        {
+            return _release_info["VERSION_ID"];
         }
+    };
 
-        // Convert the architecture value depending on distro
-        // For certain distros, use "amd64" for "x86_64"
-        // For certain distros, use "x86" for "i386"
-        if (result.architecture == "x86_64" && (
-             result.name == os::debian ||
-             result.name == os::gentoo ||
-             result.name == os::kfreebsd ||
-             result.name == os::ubuntu)) {
-            result.architecture = "amd64";
-        } else if (re_search(result.architecture, boost::regex("i[3456]86|pentium"))) {
-            // For 32-bit, use "x86" for Gentoo and "i386" for everyone else
-            if (result.name == os::gentoo) {
-                result.architecture = "x86";
-            } else {
-                result.architecture = "i386";
+    static os_linux get_os()
+    {
+        auto release_info = key_value_file(release_file::os);
+        auto const& name = release_info["NAME"];
+        if (name == "Cumulus Linux") {
+            return os_cumulus(move(release_info));
+        } else if (name == "CoreOS") {
+            return os_coreos(move(release_info));
+        }
+        return os_linux(move(release_info));
+    }
+
+    static string check_debian_linux(string const& distro_id)
+    {
+        // Check for Debian variants
+        bs::error_code ec;
+        if (is_regular_file(release_file::debian, ec)) {
+            if (distro_id == os::ubuntu || distro_id == os::linux_mint) {
+                return distro_id;
             }
+            return os::debian;
         }
+        return {};
+    }
 
-        result.selinux = collect_selinux_data();
+    static string check_oracle_linux()
+    {
+        bs::error_code ec;
+        if (is_regular_file(release_file::oracle_enterprise_linux, ec)) {
+            if (is_regular_file(release_file::oracle_vm_linux, ec)) {
+                return os::oracle_vm_linux;
+            }
+            return os::oracle_enterprise_linux;
+        }
+        return {};
+    }
 
-        return result;
+    static string check_redhat_linux()
+    {
+        bs::error_code ec;
+        if (is_regular_file(release_file::redhat, ec)) {
```

Comment:
I'm not really sure what the file format looks like. Quick search only reveals https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/5/html/Tuning_and_Optimizing_Red_Hat_Enterprise_Linux_for_Oracle_9i_and_10g_Databases/sect-Oracle_9i_and_10g_Tuning_Guide-Installing_Required_Software_Packages-Verifying_and_Updating_the_redhat_release_File.html.